test/test.hs

```diff
@@ -56,3 +64,76 @@ example_closeIsIdempotent = do
         assertFailure ("Close on a closed context threw an error : " <> show (e :: SomeException))
       Right () ->
         pure ()
+
+{-------------------------------------------------------------------------------
+  Valid IOCtxParams
+-------------------------------------------------------------------------------}
+
+-- | We test @validateIOCtxParams@ through 'withIOCtx'. If any params slip
+-- through the cracks, then the call to @setupURing@ inside 'withIOCtx' should
+-- throw an unexpected exception, which causes the property to fail.
+prop_ValidIOCtxParams :: IOCtxParams -> Property
+prop_ValidIOCtxParams params@IOCtxParams{..} =
+    checkCoverage $
+    coverTable "Result" [("Success", 5)] $
+    ioProperty $ do
+      eith <- try @IOException $ withIOCtx params $ \ctx -> pure ()
+      pure $ case eith of
+        Left e
+          |  "IOCtxParams are invalid" `isPrefixOf` ioe_location e
+            &&
+            not (
+              inBoundsExcl ioctxBatchSizeLimit batchSizeLimitBoundsExcl &&
+              inBoundsExcl ioctxConcurrencyLimit (concurrencyLimitBoundsExcl (ioctxBatchSizeLimit - 1))
+            )
+          -> tabulate "Result" [displayException e] True
+          | otherwise
+          -> counterexample ("Unknown exception: " ++ displayException e) False
+        Right () -> tabulate "Result" ["Success"] True
+  where
+    inBoundsExcl x (lb, ub) = lb < x && x < ub
+
+    batchSizeLimitBoundsExcl = (0, 2^(15 :: Int))
+    concurrencyLimitBoundsExcl batchSizeLimit =  (batchSizeLimit, 2^(16::Int))
```

Comment:
Does it really work to go up to 2^15 and 2^16? I'm surprised. I thought the limit was 4096. But I guess this test demonstrates the range.

Reply:
I was also surprised, but indeed it works